pkg/go/graph/weighted_graph_builder.go

```diff
@@ -18,60 +21,180 @@ func NewWeightedAuthorizationModelGraphBuilder() *WeightedAuthorizationModelGrap
+	wg.AddEdge(parentNode.GetUniqueLabel(), operatorNodeName, RewriteEdge, "", nil)
+	for _, child := range children {
+		err := wgb.parseRewrite(wg, operatorNode, model, child, typeDef, relation)
+		if err != nil {
+			return err
 		}
+	}
+	return nil
+}
 
-		wb.AddNode(node.uniqueLabel, node.label, node.nodeType)
+func (wgb *WeightedAuthorizationModelGraphBuilder) parseTupleToUserset(wg *WeightedAuthorizationModelGraph, parentNode *WeightedAuthorizationModelNode, model *openfgav1.AuthorizationModel, typeDef *openfgav1.TypeDefinition, rewrite *openfgav1.TupleToUserset) error {
+	// e.g. define viewer: admin from parent
+	// "parent" is the tupleset
+	tuplesetRelation := rewrite.GetTupleset().GetRelation()
+	// "admin" is the computed relation
+	computedRelation := rewrite.GetComputedUserset().GetRelation()
+
+	// find all the directly related types to the tupleset
+	relationMetadata, ok := typeDef.GetMetadata().GetRelations()[tuplesetRelation]
+	if !ok {
+		return fmt.Errorf("%w: Model cannot be parsed. %s invalid tupleset relation", ErrInvalidModel, tuplesetRelation)
 	}
+	directlyRelated := relationMetadata.GetDirectlyRelatedUserTypes()
+
+	validTTU := false
+	for _, relatedType := range directlyRelated {
+		tuplesetType := relatedType.GetType()
 
-	// Add all the edges
-	iterEdges := g.Edges()
-	for iterEdges.Next() {
-		nextEdge, ok := iterEdges.Edge().(multi.Edge)
-		if !ok {
-			return nil, fmt.Errorf("%w: could not cast %v to multi.Edge", ErrBuildingGraph, iterEdges.Edge())
+		if !typeAndRelationExists(model, tuplesetType, computedRelation) {
+			return fmt.Errorf("%w: Model cannot be parsed. %s type does not have defined %s relation", ErrInvalidModel, tuplesetType, computedRelation)
 		}
 
-		// NOTE: because we use a multigraph, one edge can include multiple lines, so we need to add each line individually.
-		iterLines := nextEdge.Lines
-		for iterLines.Next() {
-			nextLine := iterLines.Line()
-			castedEdge, ok := nextLine.(*AuthorizationModelEdge)
-			if !ok {
-				return nil, fmt.Errorf("%w: could not cast %v to AuthorizationModelEdge", ErrBuildingGraph, nextLine)
-			}
-			castedFromNode, ok := castedEdge.From().(*AuthorizationModelNode)
-			if !ok {
-				return nil, fmt.Errorf("%w: could not cast %v to AuthorizationModelNode", ErrBuildingGraph, castedEdge.From())
-			}
-			castedToNode, ok := castedEdge.To().(*AuthorizationModelNode)
-			if !ok {
-				return nil, fmt.Errorf("%w: could not cast %v to AuthorizationModelNode", ErrBuildingGraph, castedEdge.To())
-			}
-			wb.AddEdge(castedFromNode.uniqueLabel, castedToNode.uniqueLabel, castedEdge.edgeType, castedEdge.tuplesetRelation, castedEdge.conditions)
+		validTTU = true
+		rewrittenNodeName := fmt.Sprintf("%s#%s", tuplesetType, computedRelation)
+		nodeSource := wg.GetOrAddNode(rewrittenNodeName, rewrittenNodeName, SpecificTypeAndRelation)
+		typeTuplesetRelation := fmt.Sprintf("%s#%s", typeDef.GetType(), tuplesetRelation)
```

Comment:
performance: direct concatenation will be much faster than fmt.Sprintf
```suggestion
		typeTuplesetRelation := typeDef.GetType() + "#" + tuplesetRelation
```